executor/executor.go

```diff
@@ -343,6 +353,343 @@ func (e *ShowDDLJobsExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	result.Fetch(ctx)
+	return result, nil
+}
+
+type backfillResult struct {
+	nextHandle   int64
+	addedCount   int64
+	scanRowCount int64
+}
+
+func (e *RecoverIndexExec) backfillIndex(ctx context.Context) (int64, int64, error) {
+	var (
+		nextHandle    = int64(math.MinInt64)
+		totalAddedCnt = int64(0)
+		totalScanCnt  = int64(0)
+		lastLogCnt    = int64(0)
+		result        backfillResult
+	)
+	for {
+		errInTxn := kv.RunInNewTxn(e.ctx.GetStore(), true, func(txn kv.Transaction) error {
+			var err error
+			result, err = e.backfillIndexInTxn(ctx, txn, nextHandle)
+			return errors.Trace(err)
+		})
+		if errInTxn != nil {
+			return totalAddedCnt, totalScanCnt, errors.Trace(errInTxn)
+		}
+		totalAddedCnt += result.addedCount
+		totalScanCnt += result.scanRowCount
+		if totalScanCnt-lastLogCnt >= 50000 {
+			lastLogCnt = totalScanCnt
+			log.Infof("[recover-index] recover table:%v, index:%v, totalAddedCnt:%v, totalScanCnt:%v, nextHandle: %v",
+				e.table.Meta().Name.O, e.index.Meta().Name.O, totalAddedCnt, totalScanCnt, result.nextHandle)
+		}
+
+		// no more rows
+		if result.scanRowCount == 0 {
+			break
+		}
+		nextHandle = result.nextHandle
+	}
+	return totalAddedCnt, totalScanCnt, nil
+}
+
+func (e *RecoverIndexExec) fetchIdxVals(row chunk.Row, idxVals []types.Datum) []types.Datum {
+	if idxVals == nil {
+		idxVals = make([]types.Datum, 0, row.Len()-1)
+	} else {
+		idxVals = idxVals[:0]
+	}
+
+	for i := 0; i < row.Len()-1; i++ {
+		colVal := row.GetDatum(i, &e.columns[i].FieldType)
+		idxVals = append(idxVals, *colVal.Copy())
+	}
+	return idxVals
+}
+
+type recoverRows struct {
+	handle  int64
+	idxVals []types.Datum
+	skip    bool
+}
+
+func (e *RecoverIndexExec) fetchRecoverRows(ctx context.Context, srcResult distsql.SelectResult, result *backfillResult) ([]recoverRows, error) {
+	e.recoverRows = e.recoverRows[:0]
+	handleIdx := len(e.columns) - 1
+	result.scanRowCount = 0
+Loop:
+	for {
+		err := srcResult.NextChunk(ctx, e.srcChunk)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if e.srcChunk.NumRows() == 0 {
+			break
+		}
+		iter := chunk.NewIterator4Chunk(e.srcChunk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			if result.scanRowCount >= int64(e.batchSize) {
+				break Loop
+			}
+			handle := row.GetInt64(handleIdx)
+			idxVals := e.fetchIdxVals(row, e.idxValsBufs[result.scanRowCount])
+			e.idxValsBufs[result.scanRowCount] = idxVals
+			e.recoverRows = append(e.recoverRows, recoverRows{handle: handle, idxVals: idxVals, skip: false})
+			result.scanRowCount++
+			result.nextHandle = handle + 1
+		}
+	}
+
+	return e.recoverRows, nil
+}
+
+func (e *RecoverIndexExec) batchMarkDup(txn kv.Transaction, rows []recoverRows) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	sc := e.ctx.GetSessionVars().StmtCtx
+	batchDistinctKeys := make([]kv.Key, 0, len(rows))
```

Comment:
How about that we only batch get once, adding a bool slice to indicate if the key is distinct?

Reply:
Good advice.